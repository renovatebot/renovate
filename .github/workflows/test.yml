name: Test CI

on:
  push:
    branches:
      - master
      - 'renovate/**'

  pull_request:

env:
  CI: true

jobs:
  test:
    name: test
    runs-on: ${{ matrix.os }}

    # tests shouldn't need more than 15 min
    timeout-minutes: 15

    strategy:
      matrix:
        os: [ubuntu-latest, macos-latest, windows-latest]
        node-version: [10, 12]
        python-version: [3.8]
        java-version: [11]
        exclude:
          - os: windows-latest
            node-version: 10

    env:
      coverage: ${{ matrix.os == 'ubuntu-latest' && matrix.node-version == 12 && github.ref != 'refs/heads/master' }}

    steps:
      - name: Set up Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node-version }}

      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v1
        with:
          python-version: ${{ matrix.python-version }}

      - name: Set up Java ${{ matrix.java-version }}
        if: matrix.node-version == 12 && github.ref != 'refs/heads/master'
        uses: actions/setup-java@v1
        with:
          java-version: ${{ matrix.java-version }}

      - name: Skip Java tests
        if: matrix.node-version != 12 || github.ref == 'refs/heads/master'
        run: echo "::set-env name=SKIP_JAVA_TESTS::true"

      - name: Init platform
        id: init
        shell: bash
        run: |
          echo "::set-output name=yarn_cache::$(yarn cache dir)"
          python -c "from pip._internal.locations import USER_CACHE_DIR; print('::set-output name=pip_cache::' + USER_CACHE_DIR)"
          git config --global core.autocrlf false
          git config --global core.symlinks true
          git config --global user.email 'bot@renovateapp.com'
          git config --global user.name  'Renovate Bot'
          echo "Node $(node --version)"
          python --version
          echo "Yarn $(yarn --version)"
          pip --version

      - uses: actions/checkout@v2

      - name: Cache Yarn packages
        id: yarn_cache_packages
        uses: actions/cache@v1
        with:
          path: ${{ steps.init.outputs.yarn_cache }}
          key: v1-${{ runner.os }}-yarn_cache-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            v1-${{ runner.os }}-yarn_cache-

      - name: Cache node_modules
        uses: actions/cache@v1
        with:
          path: node_modules
          key: ${{ runner.os }}-${{ matrix.node-version }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-${{ matrix.node-version }}-yarn-

      - name: Cache pip packages
        id: pip_cache_packages
        uses: actions/cache@v1
        with:
          path: ${{ steps.init.outputs.pip_cache }}
          key: ${{ runner.os }}-${{ matrix.python-version }}-pip-${{ hashFiles('**/requirements.txt') }}
          restore-keys: |
            ${{ runner.os }}-${{ matrix.python-version }}-pip-

      # Clear caches on cache miss, otherwise they will grow indefinitely
      - name: Clear yarn cache
        if: steps.yarn_cache_packages.outputs.cache-hit != 'true'
        run: yarn cache clean

      - name: Clear pip cache
        if: steps.pip_cache_packages.outputs.cache-hit != 'true'
        run: python -c "from pip._internal.locations import USER_CACHE_DIR; from shutil import rmtree; rmtree(USER_CACHE_DIR, ignore_errors=True)"

      - name: Installing dependencies
        run: |
          python -m pip install --upgrade pip
          yarn install --frozen-lockfile --link-duplicates
          pip install --user -r requirements.txt

      - name: Build
        run: yarn build

      - name: Unit tests
        run: yarn jest --maxWorkers=2 --ci --coverage ${{ env.coverage }}

      - name: Upload coverage
        uses: actions/upload-artifact@v1
        if: always() && env.coverage == 'true'
        with:
          name: coverage
          path: coverage

      - name: Codecov
        shell: bash
        if: always() && env.coverage == 'true'
        continue-on-error: true
        run: bash <(curl -s https://codecov.io/bash)
        env:
          CODECOV_TOKEN: ${{ secrets.CODECOV_TOKEN }}

      - name: E2E Test
        run: yarn test-e2e

  lint:
    runs-on: ${{ matrix.os }}

    # lint shouldn't need more than 10 min
    timeout-minutes: 10

    strategy:
      matrix:
        os: [ubuntu-latest]
        node-version: [12]

    steps:
      - name: Set up Node.js ${{ matrix.node-version }}
        uses: actions/setup-node@v1
        with:
          node-version: ${{ matrix.node-version }}

      - name: Init platform
        id: init
        shell: bash
        run: |
          echo "::set-output name=yarn_cache::$(yarn cache dir)"
          git config --global core.autocrlf false
          git config --global core.symlinks true
          git config --global user.email 'bot@renovateapp.com'
          git config --global user.name  'Renovate Bot'
          echo "Node $(node --version)"
          python --version
          echo "Yarn $(yarn --version)"

      - uses: actions/checkout@v2

      - name: Cache Yarn packages
        id: yarn_cache_packages
        uses: actions/cache@v1
        with:
          path: ${{ steps.init.outputs.yarn_cache }}
          key: v1-${{ runner.os }}-yarn_cache-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            v1-${{ runner.os }}-yarn_cache-

      - name: Cache node_modules
        uses: actions/cache@v1
        with:
          path: node_modules
          key: ${{ runner.os }}-${{ matrix.node-version }}-yarn-${{ hashFiles('**/yarn.lock') }}
          restore-keys: |
            ${{ runner.os }}-${{ matrix.node-version }}-yarn-

      # Clear caches on cache miss, otherwise they will grow indefinitely
      - name: Clear yarn cache
        if: steps.yarn_cache_packages.outputs.cache-hit != 'true'
        run: yarn cache clean

      - name: Installing dependencies
        run: yarn install --frozen-lockfile --link-duplicates

      - name: Lint
        run: yarn lint

      - name: Test schema
        run: yarn test-schema

      - name: Type check
        run: yarn type-check
