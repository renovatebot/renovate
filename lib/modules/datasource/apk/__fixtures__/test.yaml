# Test fixture for APK datasource
# This represents an Alpine Linux package repository structure

# Example repository URLs that would be used:
# - https://dl-cdn.alpinelinux.org/alpine/v3.19/main
# - https://dl-cdn.alpinelinux.org/alpine/v3.19/community
# - https://packages.wolfi.dev/os

# Example package entries that would be found in APKINDEX.tar.gz:
# P:nginx
# V:1.24.0-r0
# T:nginx web server
# U:https://nginx.org/
# S:1234567
# I:1234567890
# o:nginx
# A:x86_64
# L:BSD-2-Clause
# D:openssl
# p:nginx
# c:
# r:

# Example packages that would be extracted:
packages:
  - name: nginx
    version: 1.24.0-r0
    description: nginx web server
    url: https://nginx.org/
    size: 1234567
    buildDate: 1234567890
    origin: nginx
    arch: x86_64
    license: BSD-2-Clause
    depends: [openssl]
    provides: [nginx]
    conflicts: []
    replaces: []
