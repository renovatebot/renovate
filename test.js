const fs = require('fs');

const yarnLock = fs.readFileSync('./yarn.lock').toString();
console.log(yarnLock);
/*
let yarnPackages = yarnLock.match(/[^\s]+:/ig);
yarnPackages = yarnPackages.filter(pkg => pkg.indexOf('@') !== -1);
yarnPackages = yarnPackages.map(pkg => pkg.replace(/"/g, '').replace(':', ''));
yarnPackages.forEach((entry) => {
  const atLocation = entry.indexOf('@', 1);
  const pkgName = entry.substring(0, atLocation);
  const version =
});
console.log(JSON.stringify(yarnPackages));
*/
