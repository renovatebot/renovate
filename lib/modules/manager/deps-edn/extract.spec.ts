import { Fixtures } from '../../../../test/fixtures';
import { extractPackageFile } from '.';

describe('modules/manager/deps-edn/extract', () => {
  describe('extractPackageFile', () => {
    it('returns null for invalid file', () => {
      expect(extractPackageFile('123')).toBeNull();
    });

    it('extractPackageFile', () => {
      const res = extractPackageFile(Fixtures.get('deps.edn'));
      const deps = res?.deps;
      expect(deps).toMatchObject([
        {
          depName: 'persistent-sorted-set',
          currentValue: '0.1.2',
          registryUrls: [
            'https://deps.com/foo/bar',
            'https://my.auth.com/repo',
            's3://my-bucket/maven/releases',
          ],
        },
        {
          depName: 'io.github.nextjournal/clerk',
          currentValue: '0.7.418',
          datasource: 'clojure',
        },
        { depName: 'org.clojure/clojure', currentValue: '1.9.0' },
        { depName: 'org.clojure/clojure', currentValue: '1.10.0' },
        { depName: 'org.clojure/clojurescript', currentValue: '1.10.520' },
        { depName: 'org.clojure/tools.namespace', currentValue: '0.2.11' },
        { depName: 'org.clojure/clojurescript', currentValue: '1.10.520' },
        {
          depName: 'lambdaisland/kaocha',
          packageName: 'lambdaisland/kaocha',
          currentValue: '0.0-389',
        },
        {
          depName: 'io.github.lambdaisland/kaocha-cljs',
          currentValue: '0.0-21',
        },
        {
          depName: 'lambdaisland/kaocha',
          currentValue: '0.0-389',
          depType: 'test-gitlab',
        },
        {
          depName: 'com.gitlab.lambdaisland/kaocha-cljs',
          currentValue: '0.0-21',
        },
        {
          depName: 'lambdaisland/kaocha',
          currentValue: '0.0-389',
          depType: 'test-bitbucket',
        },
        {
          depName: 'org.bitbucket.lambdaisland/kaocha-cljs',
          currentValue: '0.0-21',
        },
        { depName: 'foo/foo', currentDigest: '123', datasource: 'git-refs' },
        { depName: 'bar/bar', sourceUrl: 'https://example.com/bar' },
        { depName: 'cider/cider-nrepl', currentValue: '0.21.1' },
        { depName: 'nrepl/nrepl', currentValue: '0.6.0' },
        { depName: 'org.clojure/tools.namespace', currentValue: '0.2.11' },
        { depName: 'com.datomic/datomic-free', currentValue: '0.9.5703' },
      ]);
    });
  });
});
