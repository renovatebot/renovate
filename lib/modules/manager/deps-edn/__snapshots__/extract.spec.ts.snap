// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`modules/manager/deps-edn/extract extractPackageFile extractPackageFile 1`] = `
[
  {
    "currentValue": "0.1.2",
    "datasource": "clojure",
    "depName": "persistent-sorted-set",
    "packageName": "persistent-sorted-set:persistent-sorted-set",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version,"0.1.2"}",
  },
  {
    "currentValue": "0.7.418",
    "datasource": "clojure",
    "depName": "io.github.nextjournal/clerk",
    "packageName": "io.github.nextjournal:clerk",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version "0.7.418"}",
  },
  {
    "currentValue": "1.9.0",
    "datasource": "clojure",
    "depName": "org.clojure/clojure",
    "depType": "1.9",
    "packageName": "org.clojure:clojure",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version "1.9.0"}",
  },
  {
    "currentValue": "1.10.0",
    "datasource": "clojure",
    "depName": "org.clojure/clojure",
    "depType": "1.10",
    "packageName": "org.clojure:clojure",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version "1.10.0"}",
  },
  {
    "currentValue": "1.10.520",
    "datasource": "clojure",
    "depName": "org.clojure/clojurescript",
    "depType": "dev",
    "packageName": "org.clojure:clojurescript",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version "1.10.520"}",
  },
  {
    "currentValue": "0.2.11",
    "datasource": "clojure",
    "depName": "org.clojure/tools.namespace",
    "depType": "dev",
    "packageName": "org.clojure:tools.namespace",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version "0.2.11"}",
  },
  {
    "currentValue": "1.10.520",
    "datasource": "clojure",
    "depName": "org.clojure/clojurescript",
    "depType": "test",
    "packageName": "org.clojure:clojurescript",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version "1.10.520"}",
  },
  {
    "currentValue": "0.0-389",
    "datasource": "github-tags",
    "depName": "lambdaisland/kaocha",
    "depType": "test",
    "packageName": "lambdaisland/kaocha",
    "replaceString": "{:git/url "https://github.com/lambdaisland/kaocha.git"
                                   :git/tag "0.0-389"}",
    "sourceUrl": "https://github.com/lambdaisland/kaocha",
  },
  {
    "currentValue": "0.0-21",
    "datasource": "github-tags",
    "depName": "io.github.lambdaisland/kaocha-cljs",
    "depType": "test",
    "packageName": "lambdaisland/kaocha-cljs",
    "replaceString": "{:git/tag "0.0-21"}",
  },
  {
    "currentValue": "0.0-389",
    "datasource": "gitlab-tags",
    "depName": "lambdaisland/kaocha",
    "depType": "test-gitlab",
    "packageName": "lambdaisland/kaocha",
    "replaceString": "{:git/url "https://gitlab.com/lambdaisland/kaocha.git"
                                   :git/tag "0.0-389"}",
    "sourceUrl": "https://gitlab.com/lambdaisland/kaocha",
  },
  {
    "currentValue": "0.0-21",
    "datasource": "gitlab-tags",
    "depName": "com.gitlab.lambdaisland/kaocha-cljs",
    "depType": "test-gitlab",
    "packageName": "lambdaisland/kaocha-cljs",
    "replaceString": "{:git/tag "0.0-21"}",
  },
  {
    "currentValue": "0.0-389",
    "datasource": "gitlab-tags",
    "depName": "lambdaisland/kaocha",
    "depType": "test-bitbucket",
    "packageName": "lambdaisland/kaocha",
    "replaceString": "{:git/url "https://bitbucket.org/lambdaisland/kaocha.git"
                                   :git/tag "0.0-389"}",
    "sourceUrl": "https://bitbucket.org/lambdaisland/kaocha",
  },
  {
    "currentValue": "0.0-21",
    "datasource": "bitbucket-tags",
    "depName": "org.bitbucket.lambdaisland/kaocha-cljs",
    "depType": "test-bitbucket",
    "packageName": "lambdaisland/kaocha-cljs",
    "replaceString": "{:git/tag "0.0-21"}",
  },
  {
    "currentDigest": "123",
    "currentDigestShort": "123",
    "currentValue": null,
    "datasource": "git-refs",
    "depName": "foo/foo",
    "depType": "test-git",
    "packageName": "git@example.com/foo.git",
    "replaceString": "{:git/url "git@example.com/foo.git" :git/sha "123"}",
  },
  {
    "currentValue": null,
    "datasource": "git-refs",
    "depName": "bar/bar",
    "depType": "test-git",
    "packageName": "https://example.com/bar.git",
    "replaceString": "{:git/url "https://example.com/bar.git"}",
    "sourceUrl": "https://example.com/bar",
  },
  {
    "currentValue": "0.21.1",
    "datasource": "clojure",
    "depName": "cider/cider-nrepl",
    "depType": "repl",
    "packageName": "cider:cider-nrepl",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version "0.21.1"}",
  },
  {
    "currentValue": "0.6.0",
    "datasource": "clojure",
    "depName": "nrepl/nrepl",
    "depType": "repl",
    "packageName": "nrepl:nrepl",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version "0.6.0"}",
  },
  {
    "currentValue": "0.2.11",
    "datasource": "clojure",
    "depName": "org.clojure/tools.namespace",
    "depType": "repl",
    "packageName": "org.clojure:tools.namespace",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version "0.2.11"}",
  },
  {
    "currentValue": "0.9.5703",
    "datasource": "clojure",
    "depName": "com.datomic/datomic-free",
    "depType": "datomic",
    "packageName": "com.datomic:datomic-free",
    "registryUrls": [
      "https://deps.com/foo/bar",
      "https://my.auth.com/repo",
      "s3://my-bucket/maven/releases",
    ],
    "replaceString": "{:mvn/version  "0.9.5703"}",
  },
]
`;
